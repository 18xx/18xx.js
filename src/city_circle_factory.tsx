import { List } from 'immutable';
import * as React from 'react';
import { MouseEvent, MouseEventHandler, ReactElement } from 'react';

import CityCircle from './components/city_circle';
import Token from './components/token';

import Company from './company';
import Point from './point';

export default class CityCircleFactory {
  constructor(
    private hex: string,
    private homeTokens: List<string>,
    private onRightClickCity: Function,
    private onRightClickToken: Function,
    private tokenState: List<string>,
    private radius: number = CityCircle.DEFAULT_RADIUS,
  ) {
  }

  public build(index: number, point: Point): ReactElement<CityCircle> {
    let fn: Function;
    if (typeof this.tokenState === 'undefined' || !this.tokenState.get(index)) {
      fn = (event: MouseEvent<SVGElement>) => {
        event.preventDefault();
        this.onRightClickCity(this.hex, index);
      };
    }

    return (
      <CityCircle
        hex={this.hex}
        key={point.toString()}
        point={point}
        onContextMenu={fn}
        token={this.buildToken(index, point)}
        radius={this.radius}
      />
    );
  }

  private buildToken(index: number, point: Point): ReactElement<Token> {
    let token: ReactElement<Token>;
    let faded: boolean = false;
    let company: Company;

    const fn: MouseEventHandler<SVGElement> =
      (event: MouseEvent<SVGElement>) => {
      event.preventDefault();
      this.onRightClickToken(event, this.hex, index);
    };

    if (this.tokenState.get(index)) {
      company = Company.find(this.tokenState.get(index));
    } else if (this.homeTokens.get(index)) {
      company = Company.find(this.homeTokens.get(index));
      faded = true;
    }

    if (company) {
      token = (
        <Token
        faded={faded}
        text={company.shorthand}
        onRightClick={fn}
        primaryColor={company.primaryColor}
        radius={this.radius}
        secondaryColor={company.secondaryColor}
        textColor={company.textColor} />
      );
    }
    return token;
  }
}
