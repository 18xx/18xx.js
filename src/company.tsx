import * as Immutable from 'immutable';

interface CompanyData {
  readonly home?: string;
  readonly name: string;
  readonly shorthand: string;
  readonly primaryColor: string;
  readonly secondaryColor: string;
  readonly textColor: string;
  readonly tokens?: number;
}

class Company {
  public static fromJson(reportingMark: string, data: CompanyData): Company {
    return new Company(
      reportingMark,
      data.name,
      data.shorthand,
      data.primaryColor,
      data.secondaryColor,
      data.textColor,
    );
  }

  constructor(
    public readonly reportingMark: string,
    public readonly name: string,
    public readonly shorthand: string,
    public readonly primaryColor: string,
    public readonly secondaryColor: string,
    public readonly textColor: string
  ) {
  }
}

export default Company;
