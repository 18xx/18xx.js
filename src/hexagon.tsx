import { List } from 'immutable';

import Tile from './components/tile';

import Point from './point';

class Hexagon {
  constructor(public orientation: string | undefined) {
  }

  public get offset(): number {
    let result: number = 0;
    if (this.orientation === 'north-south') {
      result = 0.5;
    }
    return result;
  }

  public get height(): number {
    let result: number = 2 * Tile.SIDE_LENGTH;
    if (this.orientation === 'north-south') {
      result = Tile.SIDE_LENGTH * Math.sqrt(3);
    }
    return result;
  }

  public get width(): number {
    let result: number = Tile.SIDE_LENGTH * Math.sqrt(3);
    if (this.orientation === 'north-south') {
      result = 2 * Tile.SIDE_LENGTH;
    }
    return result;
  }

  public get hexTop(): number {
    return 0.25 * this.height;
  }

  public get hexBottom(): number {
    return 0.75 * this.height;
  }

  public get hexLeft(): number {
    return  0.25 * this.width;
  }

  public get hexRight(): number {
    return 0.75 * this.width;
  }

  public get center(): Point {
    return new Point(this.width / 2, this.height / 2);
  }

  public hexPoints(): List<string> {
    let result: List<string>;
    if (this.orientation === 'north-south') {
      result = List([
        `${this.hexLeft},${0}`,
        `${this.hexRight},${0}`,
        `${this.width},${this.height / 2}`,
        `${this.hexRight},${this.height}`,
        `${this.hexLeft},${this.height}`,
        `${0},${this.height / 2}`
      ]);
    } else {
      result = List([
        `${0},${this.hexTop}`,
        `${this.width / 2},${0}`,
        `${this.width},${this.hexTop}`,
        `${this.width},${this.hexBottom}`,
        `${this.width / 2},${this.height}`,
        `${0},${this.hexBottom}`
      ]);
    }
    return result;
  }
}

export default Hexagon;
