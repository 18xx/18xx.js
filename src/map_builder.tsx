import { List, Map } from 'immutable';
import * as React from 'react';
import { ReactElement } from 'react';

import CityCircle, { CityCircleProps } from './components/city_circle';
import CityName from './components/city_name';
import DynamicValues from './components/dynamic_values';
import Game from './components/game';
import MapBoard from './components/map_board';
import MapHex, {
  HEX_BOTTOM, MapHexElement, MapHexProps
} from './components/map_hex';
import OffBoard from './components/off_board';
import PrivateReservation from './components/private_reservation';
import Tile from './components/tile';
import TileCost from './components/tile_cost';
import Token from './components/token';
import Town from './components/town';

import Company from './company';
import Point from './point';
import TileBuilder from './tile_builder';
import TileDefinition from './tile_definition';
import TileSet from './tile_set';

export interface TilePromotionTuple {
  readonly hexes: List<string> | null;
  readonly promotions: List<string>;
}

// FIXME: Use proper attributes
export interface MapDefinition {
  companies: any;
  dynamicValues: any;
  hexes: any;
  impassable: any;
  names: any;
  offBoards: any;
  preplacedTile: any;
  privateReservations: any;
  tileCostTypes: any;
  tileCosts: any;
  tileManifest: any;
  tilePromotions: TilePromotionTuple[];
  towns?: any;
}

export default class MapBuilder {
  // FIXME: mapDef to interface
  constructor(
    private game: Game,
    private mapDef: MapDefinition,
    private tileSet: TileSet,
  ) {
  }

  // FIXME state is not any, and it doesn't need the whole state
  public getHexes(
    tileState: Map<string, string>,
    tokenState: Map<string, List<string>>,
  ): List<ReactElement<MapHex>> {
    const keys: string[] = Object.keys(this.mapDef.hexes);
    // FIXME: hexData to interface
    const hexData: MapHexProps[] = [];

    for (const row of keys) {
      const columns: number[] = this.mapDef.hexes[row];

      for (const column of columns) {
        let fill: string;
        const hexElements: MapHexElement[] = [];
        const hex: string = row + column;

        if (this.mapDef.names[hex]) {
          let y: number = HEX_BOTTOM + 4;
          if (this.mapDef.offBoards[hex]) {
            y -= 20;
          }

          hexElements.push(
            <CityName
              key='city-name'
              point={new Point(Tile.CENTER.x, y)}
              name={this.mapDef.names[hex]} />
          );
        }

        if (this.mapDef.names[hex] && !this.mapDef.offBoards[hex]) {
          let ccProps: CityCircleProps = { point: Tile.CENTER, key: 'cc' };
          const companyStr: string = Map<string, any>(
            this.mapDef.companies
          ).findKey(c => c.home === hex);

          if (companyStr) {
            const company: Company = Company.find(companyStr);
            const token: ReactElement<Token> = (
              <Token
              faded={true}
              text={company.shorthand}
              primaryColor={company.primaryColor}
              secondaryColor={company.secondaryColor}
              textColor={company.textColor} />
            );
            ccProps = {
              ...ccProps,
              token
            };
          }
          hexElements.push(React.createElement(CityCircle, ccProps));
        }

        if (this.mapDef.towns && this.mapDef.towns[hex]) {
          if (this.mapDef.towns[hex] === 1) {
            hexElements.push(
              <Town points={List<Point>([Tile.CENTER])} key='town' />
            );
          } else {
            const points: List<Point> = List([
              new Point(Tile.CENTER.x - 20, Tile.CENTER.y + 10),
              new Point(Tile.CENTER.x + 20, Tile.CENTER.y - 10),
            ]);
            hexElements.push(
              <Town points={points} key='map-towns' />
            );
          }
        }

        if (this.mapDef.offBoards[hex]) {
          fill = 'red';
          const exits: List<number> = List<number>(
            this.mapDef.offBoards[hex].exits as string[]
          );
          hexElements.push(
            // FIXME: Missing type for offboard exits
            <OffBoard
              key='off-board'
              exits={exits} />
          );
        }

        if (this.mapDef.dynamicValues[hex]) {
          hexElements.push(
            <DynamicValues
              key='dv'
              values={this.mapDef.dynamicValues[hex].values}
              fixedHeight={this.mapDef.dynamicValues[hex].fixedHeight} />
          );
        }

        if (this.mapDef.tileCosts[hex]) {
          hexElements.push(this.assignTileCost(hex));
        }

        let homeTokens: List<string> = List<string>([]);
        const companyStr: string = Map<string, any>(
          this.mapDef.companies
        ).findKey(
          c => c.home === hex
        );

        if (companyStr) {
          homeTokens = List<string>([companyStr]);
        }

        let tile: ReactElement<Tile>;
        const tileBuilder: TileBuilder = new TileBuilder(
          this.game.onRightClickCity,
          hex,
          tokenState.get(hex),
          homeTokens,
        );

        if (tileState.has(hex)) {
          const tileStr: string[] = tileState.get(hex).split('.');
          tile = tileBuilder.buildTile(
            this.tileSet.findDefinition(tileStr[0]),
            parseInt(tileStr[1], 10)
          );
        } else if (this.mapDef.preplacedTile[hex]) {
          tile = tileBuilder.buildTile(
            new TileDefinition(this.mapDef.preplacedTile[hex])
          );
        }

        if (this.mapDef.privateReservations[hex]) {
          const name: string = this.mapDef.privateReservations[hex];
          hexElements.push(
            <PrivateReservation key='pcr' name={name} />
          );
        }

        hexData.push({
          row,
          column,
          fill,
          tile,
          elements: List(hexElements)
        });
      }
    }

    return this.renderHexes(hexData);
  }

  public get addOnTop(): List<ReactElement<any>> {
    if (this.mapDef.impassable) {
      return List<ReactElement<any>>(this.mapDef.impassable.map(
        (hexes: [string, string]) => {
          const hex0: MapHex = new MapHex({
            column: parseInt(hexes[0].substring(1), 10),
            row: (hexes[0].substring(0, 1)),
          });

          const hex1: MapHex = new MapHex({
            column: parseInt(hexes[1].substring(1), 10),
            row: (hexes[1].substring(0, 1)),
          });

          let x1: number;
          let x2: number;
          if (hex0.absoluteCenter.y === hex1.absoluteCenter.y) {
            x1 = x2 = (hex0.absoluteCenter.x + hex1.absoluteCenter.x) / 2;
          } else {
            x1 = hex1.absoluteCenter.x;
            x2 = hex0.absoluteCenter.x;
          }

          let y1: number;
          if (hex0.absoluteCenter.y > hex1.absoluteCenter.y) {
            y1 = Tile.SIDE_LENGTH / -2;
          } else {
            y1 = Tile.SIDE_LENGTH / 2;
          }
          const y2: number = -y1;

          return (
            <line
            key={`impassable-${hexes.join('-')}`}
            x1={x1}
            y1={hex0.absoluteCenter.y + y1}
            x2={x2}
            y2={hex1.absoluteCenter.y + y2}
            strokeWidth={4}
            stroke='red' />
          );
        }
      ));
    }
  }

  private assignTileCost(hex: string): ReactElement<TileCost> {
    // FIXME interface
    const type: any = this.mapDef.tileCosts[hex];
    // FIXME interface
    const costs: any = this.mapDef.tileCostTypes[type];

    return (
      <TileCost
      amount={costs.amount}
      color={costs.color}
      key='tile-cost'
      shape={costs.shape} />
    );
  }

  // FIXME: Typedef
  private renderHexes(hexData: MapHexProps[]): List<ReactElement<MapHex>> {
    return List(hexData.map(data =>
      <MapHex
        key={data.row + data.column}
        row={data.row}
        column={data.column}
        fill={data.fill}
        tile={data.tile}
        elements={List(data.elements)}
        onHexClick={this.game.onHexClick}
      />
    ));
  }
}
