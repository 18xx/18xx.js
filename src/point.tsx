import Tile from './components/tile';

export interface PointDefinition {
  readonly x: number;
  readonly y: number;
}

export default class Point {
  public static fromCenter(radians: number, distance?: number): Point {
    if (!distance) {
      distance = Tile.WIDTH / 2;
    }

    return new Point(
      Tile.CENTER.x + Math.cos(radians * Math.PI / 3) * distance,
      Tile.CENTER.y - Math.sin(radians * Math.PI / 3) * distance
    );
  }

  constructor(public x: number, public y: number) {
    this.x = x;
    this.y = y;
  }

  public toString(): string {
    return `${this.x},${this.y}`;
  }

  public isAtCenterX(): boolean {
    return (Math.abs(this.x - Tile.CENTER.x) < 0.001);
  }
}
