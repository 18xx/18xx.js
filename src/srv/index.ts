import * as bodyParser from 'body-parser';
import * as crypto from 'crypto';
import * as express from 'express';
import * as redis from 'redis';

const app: express.Express = express();
app.use(express.static('dist'));
app.use(bodyParser.json());

const redisClient: redis.RedisClient = redis.createClient();

const html: Function = (game: string, initialState?: string) => `
<html>
<head>
  <link rel='stylesheet' type='text/css' href='/styles.css'>
  <div id='container'
    data-game-name='${game}'
    data-initial-state-id='${initialState}'>
  </div>
  <script src="/build.js"></script>
</head>
<body></body>
</html>
`;

app.get('/maps/:game', (req, res) => {
  res.send(html(req.params.game));
});

app.get('/maps/:game/:state', (req, res) => {
  res.send(html(req.params.game, req.params.state));
});

app.get('/state/:state', (req, res) => {
  redisClient.get(req.params.state, (err: any, reply: any) => {
    res.json(JSON.parse(reply));
  });
});

app.post('/update', (req, res) => {
  const jsonBody: string = JSON.stringify(req.body);
  const hash: string = crypto.createHash('md5').update(jsonBody).digest('hex');
  redisClient.set(hash, jsonBody);
  res.send(true);
});

app.listen(3002);
