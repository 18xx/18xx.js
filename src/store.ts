import { GameState } from './reducers/game';

export interface Store {
  getState: (hash: string) => Promise<GameState>;
  setState: (hash: string, body: GameState) => Promise<boolean>;
}
