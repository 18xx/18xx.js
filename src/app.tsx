import * as Immutable from 'immutable';
import { List } from 'immutable';
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import { createStore, Store } from 'redux';
import '../css/all.css';

import AllTiles from './components/all_tiles';
import { HistoryEntry } from './components/history';
import Game from './containers/game';

import { MapDefinition } from './map_builder';
import { TileDefinitionInput } from './tile_definition';

import { game, GameState, initialState } from './reducers/game';

import * as mapDef1817 from '../config/maps/1817.json';
import * as mapDef1830 from '../config/maps/1830.json';
import * as mapDef1849 from '../config/maps/1849.json';
import * as mapDef1880 from '../config/maps/1880.json';
import * as mapDef1889 from '../config/maps/1889.json';
let mapDef: MapDefinition;

import * as allTilesJson from '../config/tiles.json';

const allTiles: List<TileDefinitionInput> =
  Immutable.fromJS(allTilesJson).map(
    (el: any) => el.toJS() as TileDefinitionInput
  );

const container: HTMLElement | null = document.getElementById('map-container');
if (!container) {
  throw new Error('Did not find map-container element');
}

const initialStateId: string | undefined = container.dataset.initialStateId;

let init: (state: any) => void;
if (container.dataset.gameName && container.dataset.gameName !== 'tiles') {
  switch (container.dataset.gameName) {
    case '1817':
      mapDef = mapDef1817 as any;
      break;
    case '1830':
      mapDef = mapDef1830 as any;
      break;
    case '1849':
      mapDef = mapDef1849 as any;
      break;
    case '1880':
      mapDef = mapDef1880 as any;
      break;
    case '1889':
      mapDef = mapDef1889 as any;
      break;
    default:
      throw new Error(`Unsupported Game: ${container.dataset.gameName}`);
  }

  init = (state: GameState): void => {
    const store: Store<GameState> = createStore(
      game,
      state,
      (window as any).__REDUX_DEVTOOLS_EXTENSION__ &&
        (window as any).__REDUX_DEVTOOLS_EXTENSION__()
    );

    ReactDOM.render(
      <Provider store={store}>
        <Game gameName={container.dataset.gameName!} mapDef={mapDef} />
      </Provider>,
      container
    );
  };
} else {
  init = (state: GameState): void => {
    const store: Store<GameState> = createStore(
      game,
      state,
      (window as any).__REDUX_DEVTOOLS_EXTENSION__ &&
        (window as any).__REDUX_DEVTOOLS_EXTENSION__()
    );

    ReactDOM.render(
      <Provider store={store}>
        <AllTiles mapDef={mapDef1830 as any} tiles={allTiles} />
      </Provider>,
      container
    );
  };
}

if (initialStateId === 'undefined') {
  init({
    ...initialState,
    name: container.dataset.gameName,
  });
} else {
  fetch('/state/' + initialStateId).then(res => {
    res.json().then(data => {
      // FIXME: Don't hardcode tiles & tokens
      const newState: GameState = {
        history: Immutable.List<HistoryEntry>(data.history),
        name: container.dataset.gameName!,
        tiles: Immutable.fromJS(data.tiles),
        tokens: Immutable.fromJS(data.tokens),
      };
      init(newState);
    });
  });
}
