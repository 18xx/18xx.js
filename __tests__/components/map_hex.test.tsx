import { List } from 'immutable';

import * as React from 'react';
import { ReactElement } from 'react';
import * as renderer from 'react-test-renderer';

import Hexagon from '../../src/hexagon';
import { MapDefinition } from '../../src/map_builder';

import MapHex from '../../src/components/map_hex';
import Tile from '../../src/components/tile';

describe('MapHex', () => {
  const mapDef: MapDefinition = {
    companies: {},
    orientation: 'east-west',
    tileManifest: {},
  };

  describe('#absoluteLeft', () => {
    describe('when column is one', () => {
      it('returns zero', () => {
        const subject: MapHex = new MapHex({ mapDef, row: 'a', column: 1 });
        expect(subject.absoluteLeft).toEqual(0);
      });
    });

    describe('when column is 6', () => {
      it('returns 277.5', () => {
        const subject: MapHex = new MapHex({ mapDef, row: 'a', column: 6 });
        expect(subject.absoluteLeft).toBeCloseTo(277.128);
      });
    });
  });

  describe('#absoluteTop', () => {
    describe('when row is a', () => {
      it('returns zero', () => {
        const subject: MapHex = new MapHex({ mapDef, row: 'a', column: 1 });
        expect(subject.absoluteTop).toEqual(0);
      });
    });

    describe('when row is g', () => {
      it('returns something else', () => {
        const subject: MapHex = new MapHex({ mapDef, row: 'g', column: 6 });
        expect(subject.absoluteTop).toEqual(576);
      });
    });
  });

  describe('#hex', () => {
    it('returns the row and column', () => {
      const subject: MapHex = new MapHex({ mapDef, row: 'u', column: 46 });
      expect(subject.hex).toEqual('u46');
    });
  });

  describe('#toString()', () => {
    it('returns the SVG for this element', () => {
      const subject: any = renderer.create(
        <MapHex mapDef={mapDef} row='e' column={11} />
      );
      expect(subject).toMatchSnapshot();
    });

    describe('when it contains a tile', () => {
      it('draws the tile at the end', () => {
        const tile: ReactElement<Tile> = (
          <Tile color='yellow' orientation='east-west' />
        );
        const subject: any = renderer.create(
          <MapHex mapDef={mapDef} row='e' column={11} tile={tile} />
        );
        expect(subject).toMatchSnapshot();
      });
    });

    describe('when it contains some elements', () => {
      it('draws the tile at the end', () => {
        const elements: List<string> = List(['ELEMENT']);
        const subject: any = renderer.create(
          <MapHex mapDef={mapDef} row='e' column={11} elements={elements} />
        );
        expect(subject).toMatchSnapshot();
      });
    });
  });
});
