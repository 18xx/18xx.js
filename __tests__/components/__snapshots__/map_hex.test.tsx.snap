exports[`MapHex #toString() returns the SVG for this element 1`] = `
<svg
  className="hex"
  onClick={[Function]}
  x={554.2562584220407}
  y={384}>
  <polygon
    fill="#efe"
    points="0,32 55.42562584220407,0 110.85125168440814,32 110.85125168440814,96 55.42562584220407,128 0,96"
    stroke="black" />
</svg>
`;

exports[`MapHex #toString() when it contains a tile draws the tile at the end 1`] = `
<svg
  className="hex"
  onClick={[Function]}
  x={554.2562584220407}
  y={384}>
  <polygon
    fill="#efe"
    points="0,32 55.42562584220407,0 110.85125168440814,32 110.85125168440814,96 55.42562584220407,128 0,96"
    stroke="black" />
  <svg
    height={128}
    width={110.85125168440814}
    xmlns="http://www.w3.org/2000/svg">
    <polygon
      fill="yellow"
      points="55.42562584220407,0 110.85125168440814,32 110.85125168440814,96 55.42562584220407,128 0,96 0,32" />
  </svg>
</svg>
`;

exports[`MapHex #toString() when it contains some elements draws the tile at the end 1`] = `
<svg
  className="hex"
  onClick={[Function]}
  x={554.2562584220407}
  y={384}>
  <polygon
    fill="#efe"
    points="0,32 55.42562584220407,0 110.85125168440814,32 110.85125168440814,96 55.42562584220407,128 0,96"
    stroke="black" />
  ELEMENT
</svg>
`;
